Lp batch experiment

minchopaskal opened the pull request:
Use lpBatchAppend/Insert instead of multiple calls to lpAppend/Insert wherever possible

# Hash

```
$ memtier_benchmark --test-time 60 --distinct-client-seed "--data-size" "64" --command="DEL key" --command "HSET key field:1 __data__ field:2 __data__ field:3 __data__ field:4 __data__ field:5 __data__ field:6 __data__ field:7 __data__ field:8 __data__ field:9 __data__ field:10 __data__ field:11 __data__ field:12 __data__ field:13 __data__ field:14 __data__ field:15 __data__ field:16 __data__ field:17 __data__ field:18 __data__ field:19 __data__ field:20 __data__ field:21 __data__ field:22 __data__ field:23 __data__ field:24 __data__ field:25 __data__ field:26 __data__ field:27 __data__ field:28 __data__ field:29 __data__ field:30 __data__ field:31 __data__ field:32 __data__ field:33 __data__ field:34 __data__ field:35 __data__ field:36 __data__ field:37 __data__ field:38 __data__ field:39 __data__ field:40 __data__ field:41 __data__ field:42 __data__ field:43 __data__ field:44 __data__ field:45 __data__ field:46 __data__ field:47 __data__ field:48 __data__ field:49 __data__ field:50 __data__" -c 50 -t 4 --hide-histogram 

========================================================================================================
Type                Ops/sec    Avg. Latency     p50 Latency     p99 Latency   p99.9 Latency       KB/sec
--------------------------------------------------------------------------------------------------------
Hsets-before       49539.14         2.03172         2.03900         3.03900         3.79100    206598.42
Hsets-after        50346.67         1.99950         1.99900         3.03900         3.85500    209966.12
```

# stream

```
memtier_benchmark "--data-size" "100" --command "XADD __key__ MAXLEN ~ 5 * field __data__ field __data__ field __data__ field __data__ field __data__" --command-key-pattern="P" --key-minimum=1 --key-maximum 1000000 --test-time 60 -c 50 -t 4 --hide-histogram
```

## no pipelining

```
==================================================================================================
Type              Ops/sec    Avg. Latency     p50 Latency     p99 Latency   p99.9 Latency       KB/sec
--------------------------------------------------------------------------------------------------
Xadds-before      168635.18         1.18610         1.18300         1.93500         2.51100    112954.10
Xadds-after       167506.27         1.19409         1.19100         1.94300         2.51100    112197.94
```

## --pipeline 5

```
==================================================================================================
Type              Ops/sec    Avg. Latency     p50 Latency     p99 Latency   p99.9 Latency       KB/sec
--------------------------------------------------------------------------------------------------
Xadds-before      400853.08         2.49440         2.43100         3.95100         5.18300    268496.78
Xadds-after       393134.24         2.54335         2.47900         3.98300         4.57500    263326.57
```

## --pipeline 10

```
==================================================================================================
Type              Ops/sec    Avg. Latency     p50 Latency     p99 Latency   p99.9 Latency       KB/sec
--------------------------------------------------------------------------------------------------
Xadds-before      467683.38         4.27567         4.19100         6.04700         7.48700    313260.56
Xadds-after       456249.25         4.38281         4.31900         6.30300         7.26300    305601.83
```

# zset

Show a little bit of improvement for the insert case and seemingly little difference for the append case.

## Insert 

### Script

```
./src/redis-cli flushall
./src/redis-cli zadd key 100000 a 200000 b 300000 c 400000 d 500000 e 600000 f
./src/redis-cli -p 5002 flushall
./src/redis-cli -p 5002 zadd key 100000 a 200000 b 300000 c 400000 d 500000 e 600000 f

echo UNSTABLE: && memtier_benchmark --command="zadd key __key__ c" --key-prefix="" --command-key-pattern=S --hide-histogram --key-minimum=300001 --test-time 180 --pipeline 10 | tail -n 2 | head -n 1
echo BATCH: && memtier_benchmark --command="zadd key __key__ c" --key-prefix="" --command-key-pattern=S --hide-histogram --key-minimum=300001 --test-time 180 --pipeline 10 --port 5002 | tail -n 2 | head -n 1
``` 

### Results

```
==================================================================================================
Type              Ops/sec    Avg. Latency     p50 Latency     p99 Latency   p99.9 Latency       KB/sec
--------------------------------------------------------------------------------------------------
Zadds-before      1249850.63         1.59968         1.59100         2.70300         3.39100     56606.65
Zadds-after       1287818.39         1.55250         1.53500         2.71900         3.31100     58349.31
```

## Append

### Script

```
./src/redis-cli flushall
./src/redis-cli zadd key 100000 a 200000 b 300000 c 400000 d 500000 e 600000 f
./src/redis-cli -p 5002 flushall
./src/redis-cli -p 5002 zadd key 100000 a 200000 b 300000 c 400000 d 500000 e 600000 f

echo UNSTABLE: && memtier_benchmark --command="zadd key __key__ c" --key-prefix="" --command-key-pattern=S --hide-histogram --key-minimum=600001 --test-time 180 --pipeline 10 | tail -n 2 | head -n 1
echo BATCH: && memtier_benchmark --command="zadd key __key__ c" --key-prefix="" --command-key-pattern=S --hide-histogram --key-minimum=600001 --test-time 180 --pipeline 10 --port 5002 | tail -n 2 | head -n 1
```

### Results

```
==================================================================================================
Type               Ops/sec    Avg. Latency     p50 Latency     p99 Latency   p99.9 Latency       KB/sec
--------------------------------------------------------------------------------------------------
Zadds-before       1249664.05         1.59991         1.58300         2.81500         3.40700     56923.61
Zadds-after        1258994.60         1.58803         1.56700         2.86300         3.37500     57351.86
```

## Append optimization ( **REVERTED** )

Managed to optimize the append case (updating the score of the biggest element with a bigger score) with commit [6036cc7](https://github.com/minchopaskal/redis/pull/4/commits/6036cc701dfa74925726b71efe442fb4cc837532)

### Script

The same as above

### Results

```
==================================================================================================
Type               Ops/sec    Avg. Latency     p50 Latency     p99 Latency   p99.9 Latency       KB/sec
--------------------------------------------------------------------------------------------------
append-before       1240420.18         1.61181         1.59100         2.86300         3.34300     55891.71
append-after        1396518.14         1.43161         1.39100         2.65500         3.08700     63056.35
```

minchopaskal committed:
Optimize hashTypeSet with lpBatchAppend

minchopaskal committed:
Improve streamAppendItem with lpBatchAppen

minchopaskal committed:
Update t_zset with lpBatchAppend

minchopaskal committed:
fix memleak in t_stream.c

tezc commented on src/t_stream.c:
```diff
@@ -548,15 +548,25 @@ int streamAppendItem(stream *s, robj **argv, int64_t numfields, streamID *added_
         if (server.stream_node_max_bytes > 0 && server.stream_node_max_bytes < prealloc) {
             prealloc = server.stream_node_max_bytes;
         }
-        lp = lpNew(prealloc);
```
@sundb maybe you can comment on this experiment. Remember we added lpBatchInsert() a while ago. So, we wonder if we can use it on some other places and gain some performance benefit. There is one scenario on hset and another on zset. Potentially we can see some benefit on these places but so far on benchmarks we did not see any difference. 

Stream case is different. I see that we preallocate 4kb for each key. Because appending multiple elements is slow. So maybe we can use lpBatchInsert() to make inserts faster. Then, we don't have to preallocate 4kb. On average, it will be like using a few kbs less per key. Does that make sense? Do you think is it worth it?  I don't know streams much, maybe you can comment on this.

sundb replied:
i think `lpAppend` is actually quite fast, because it doesn't need to move the memory in the tail.
the reason why zset has a benefit is that it can save the operation of moving memory from insertion.

sundb replied:
Another optimization is similar to lpNextBatch, which uses `lpSkip` to move and verify at the end.
Because in the stream, we might skip a bunch of entries via `lpNext`, but I tried it, and didn't see any benefit.
like:
```
        p = lpNext(lp,p); /* Skip the first field. */
        for (int64_t j = 0; j < master_fields_count; j++)
            p = lpNext(lp,p); /* Skip all master fields. */
        p = lpNext(lp,p); /* Skip the zero master entry terminator. */
```

minchopaskal replied:
@tezc I spend some time on deciphering the inner workings of the stream and I think I can answer your question.

Generally the stream is stored in a radix tree. 
It's keys are stream ids (i.e 128-bit number representing timestamp+sequence number). For each key we have a listpack of stream fields+values. 
The listpack contains two sections:
- master entry - contains a base stream id + count of fields + the names of fields in the stream
- flag + delta encoded stream id(relative to the master id) + list of field-value pairs

The listpack is by default 4KB big or 100 fields big (whichever comes first). Both can be controlled by the config.

So whenever we add a new entry in the stream if there is enough space for it in the current latest listpack (i.e the one with the biggerst stream id) then this entry is added to that listpack. (there is also the detail that if the new entry's fields are exactly the same as the fields in the master entry section, then we don't write the fields inside the listpack - this is controlled by a flag)
If it's not big enough *it's shrank to fit* and a new listpack is created. The stream-id for the new entry becomes the master id for that listpack. The listpack is created using prealloc of 4KB (**NOTE** that if the configurations specifies less max-memory for the stream the prealloc will use that).

Given all these details I think the preallocation is used as optimally as possible - we never preallocate more than we can ever use so we guarantee no reallocs will ever happen when we add to a listpack. Also whenever we add too much (or too big in size) entries we are guaranteed no extra space is wasted because we call `shrinkToFit` on the listpack before creating a new one.

Another detail is `streamTrim`. If `MAXLEN` option is given to `XADD` then some elements of the added stream entry may be deleted. The deletion though happens with flags so the memory is still used. So still no downsides to prealloc.

This research gave me ideas on how to properly test if `lpBatch*` actually gives better results. Will post my findings later.

EDIT: fixed incorrect info about default max elements.
EDIT 2: added info about `streamTrim`

tezc replied:
Overall, I don't expect performance gain on stream case. I thought, the reason for preallocation is there are bunch of appends which are slow due to realloc on each lpAppend(). So, an idea, if we don't use preallocation and use lpBatchAppend(), are we going to see the same performance numbers? Then, perhaps we don't have to do preallocation and save a few KBs per key, no?

minchopaskal replied:
No, I was not talking about performance but exactly about the benefit or preallocation - batch or no batch prealloc is calculated to be as optimal as possible - f.e if default size if 4KB, then we will put elements inside the listpack as much as possible. Never putting more elements crosses the 4KB boundry we will shrink to fit, so no memory is wasted. NOTE that whenever we are putting data into a stream we are working on 1 and only 1 listpack at a time, so even if applicable in certain cases we will save at most 4KB.

`prealloc(0) + lpBatch` is pessimization on performance because that will certainly introduce reallocs and as described in this comment - at most an insignificant memory optimization. Here is a simple case why reallocs will happend:
- we insert a new entry of size 1KB - listpack of size 1KB is allocated.
- after that we insert another entry of 1KB - it will be written inside the same listpack but it'll need to be reallocated to 2KB.
- etc.

Batching will make reallocations happen less freqently ofcourse, but still the code will try to push as much elements as possible until reaching the max-memory mark.

> Overall, I don't expect performance gain on stream case

Mhm, I think so too. But I'm trying to benchmark it still, maybe we can squeeze something out of it :)

minchopaskal commented:
Updated top comment with new results for zset. Actually it shows a bit of non-trivial improvement for the `insert` case and a little difference for the `append` case.

EDIT: Added new optimization for the append case of zsetAdd (unrelated to lpBatch)

minchopaskal committed:
Optimize zadd when replacing biggest element with bigger score

sundb commented on src/t_zset.c:
```diff
@@ -1452,8 +1452,29 @@ int zsetAdd(robj *zobj, double score, sds ele, int in_flags, int *out_flags, dou
             if (score != curscore) {
-                zobj->ptr = zzlDelete(zobj->ptr,eptr);
-                zobj->ptr = zzlInsert(zobj->ptr,ele,score);
+                unsigned char *sptr = lpNext(zobj->ptr, eptr);
+
+                /* If we are updating the score of the last element(i.e the one
+                 * with the biggest score) with bigger score, then it's 
+                 * enough to just replace its score. This will preserve the
+                 * ordering and save us some iterations over the listpack. */
+                int is_last = (lpNext(zobj->ptr, sptr) == NULL);
+                if (is_last && score > curscore) {
+                    char scorebuf[MAX_D2STRING_CHARS];
+                    int scorelen = 0;
+                    long long lscore;
+                    int score_is_long = double2ll(score, &lscore);
+                    if (!score_is_long)
+                        scorelen = d2string(scorebuf,sizeof(scorebuf),score);
+
+                    if (score_is_long)
+                        zobj->ptr = lpReplaceInteger(zobj->ptr, &sptr, lscore);
+                    else
+                        zobj->ptr = lpReplace(zobj->ptr, &sptr, (unsigned char*)scorebuf, scorelen);
+                } else {
+                    zobj->ptr = zzlDelete(zobj->ptr,eptr);
+                    zobj->ptr = zzlInsert(zobj->ptr,ele,score);
+                }
```
How much benefit can this optimization bring? code become ugly.

minchopaskal replied:
I think this can be in separate function because it duplicates code in zzlInsertAt. 
Improvement is ~13% (I've edited the top comment to show the difference)

tezc replied:
I'm thinking out loud and assuming random operations. Say we have 10 elements. Chance of zadd being called for the last element is ~%10 (if it is not an addition). Though, half of the time score will be less, then can we say it will speed up %5 of the (update) operations. If we have more elements, benefit will diminish. I'm not sure if it is worth the extra code. Because it feels like we are optimizing for a very specific case only (hope the above reasoning is correct).

minchopaskal replied:
I agree with both statements - the code indeed becomes ugly and it's questionable if the change is needed. Only reference I found is this - https://github.com/redis/redis/issues/5179 when the same optimization was added for the skiplist case of zset, but the code is way more cleaner there.

Will revert.

minchopaskal committed:
Revert "Optimize zadd when replacing biggest element with bigger score"

This reverts commit 6036cc701dfa74925726b71efe442fb4cc837532.

minchopaskal committed:
Revert use of lpBatchInsert/Append in stream

minchopaskal commented:
Ultimately decided to revert changes in stream with [fc6d407](https://github.com/minchopaskal/redis/pull/4/commits/fc6d407a9bb44b93c6d1d53e89d628fd65557ddf) after following benchmark:

## Script
```
./src/redis-cli flushall
./src/redis-cli -p 5002 flushall

echo UNSTABLE: && memtier_benchmark "--data-size" "512" --command "XADD key * field __data__ field __data__ field __data__ field __data__ field __data__" -c 50 -t 4 --hide-histogram --test-time 75 --pipeline 10 | tail -n 2 | head -n 1
echo BATCH: && memtier_benchmark "--data-size" "512" --command "XADD key * field __data__ field __data__ field __data__ field __data__ field __data__" -c 50 -t 4 --hide-histogram --test-time 75 --pipeline 10 --port 5002 | tail -n 2 | head -n 1
```

## Results
```
==================================================================================================
Type              Ops/sec    Avg. Latency     p50 Latency     p99 Latency   p99.9 Latency       KB/sec
--------------------------------------------------------------------------------------------------
Xadds-before      371864.85         5.37769         5.75900         6.62300         7.74300    984027.16
Xadds-after       357807.08         5.58897         5.98300         6.84700         7.83900    946823.47
```
